examples/trusted_information_retrieval/backend/src/main.rs

```diff
@@ -0,0 +1,142 @@
+    Request, Response, Status,
+};
+
+#[derive(StructOpt, Clone)]
+#[structopt(about = "Aggregator Backend")]
+pub struct Opt {
+    #[structopt(long, help = "Private RSA key file used by gRPC server")]
+    grpc_tls_private_key: String,
+    #[structopt(
+        long,
+        help = "PEM encoded X.509 TLS certificate file used by gRPC server"
+    )]
+    grpc_tls_certificate: String,
+    #[structopt(
+        long,
+        help = "Database URL",
+        default_value = "https://tfl.gov.uk/tfl/syndication/feeds/cycle-hire/livecyclehireupdates.xml"
+    )]
+    database_url: String,
+}
+
+pub struct DatabaseService {
+    entries: Arc<Mutex<Vec<DatabaseEntry>>>,
+}
+
+#[tonic::async_trait]
+impl Database for DatabaseService {
+    async fn get_database_size(
+        &self,
+        _request: Request<GetDatabaseSizeRequest>,
+    ) -> Result<Response<GetDatabaseSizeResponse>, Status> {
+        let entries_lock = self.entries.lock().await;
+        Ok(Response::new(GetDatabaseSizeResponse {
+            size: entries_lock.len() as i32,
+        }))
+    }
+
+    async fn list_database_entries(
+        &self,
+        request: Request<ListDatabaseEntriesRequest>,
+    ) -> Result<Response<ListDatabaseEntriesResponse>, Status> {
+        let first = request.get_ref().first_entry as usize;
+        let last = request.get_ref().last_entry as usize;
+        let entries_lock = self.entries.lock().await;
+        let response = ListDatabaseEntriesResponse {
+            entries: entries_lock[first..last].to_vec(),
+        };
+        Ok(Response::new(response))
+    }
+}
+
+/// Loop for requesting a new database version from `url`.
+async fn update_database_loop(url: Url, database: Arc<Mutex<Vec<DatabaseEntry>>>) {
+    loop {
+        info!("Updating database");
+        match load_database(url.clone()).await {
+            Ok(updated_database) => {
+                let mut database_lock = database.lock().await;
+                *database_lock = updated_database;
+            }
+            Err(error) => error!("Error loading database: {:?}", error),
+        }
+
+        // Wait for a minute before updating the database.
+        tokio::time::delay_for(tokio::time::Duration::from_secs(60)).await;
+    }
+}
+
+#[tokio::main]
+async fn main() -> Result<(), Box<dyn std::error::Error>> {
+    env_logger::init();
+    let opt = Opt::from_args();
+    info!("Running backend database");
+
+    let private_key = tokio::fs::read(&opt.grpc_tls_private_key).await?;
+    let certificate = tokio::fs::read(&opt.grpc_tls_certificate).await?;
+    let database_url = Url::parse(&opt.database_url).expect("Couldn't parse database URL");
+
+    // Create a mutex-protected database.
+    let database = Arc::new(Mutex::new(vec![]));
+
+    // Run a separate thread for updating the database.
+    tokio::spawn(update_database_loop(database_url, database.clone()));
+
+    let handler = DatabaseService { entries: database };
```

Comment:
It would require copying `DatabaseService` into a separate asynchronous task or using it as an `Arc<Mutex<...`.
And `entries` inside `DatabaseService` are already an `Arc<Mutex<...`. So it would add lock inside a lock.